### ISSUE TITLE
Session Expired when call user Authenticate API WebService

### ISSUE THREAD

**Issue by LeCaoPhuc** [odoo/odoo#29449]
Impacted versions: 12

Steps to reproduce: 
Open browser in my odoo page.
Call Ajax with API authenticate: 
```   
$.ajax({
    url: 'https://your_url/web/session/authenticate',
    headers: {
        'Content-Type':'application/json',
    },
    method: 'POST',
    data: JSON.stringify({
 		jsonrpc: "2.0",
            	method: 'call',
		params : {
                    db:  'dbName'
                    login: 'username',
                    password: 'password'
                }
            }),
    success: function(data){
      console.log('succes: '+data);
    },
	error: function(error) {
	console.log(error)
}
  });
```
I got ```session_id``` in result.
After then I use ```session_id``` to call API :
```
$.ajax({
    url: 'https://your_url/web/dataset/call_kw/res.users/read',
    headers: {
        'Content-Type':'application/json',
	"X-Openerp-Session-Id": 'session_id_in_before_authenticate_API'
    },
    method: 'POST',
    data: JSON.stringify({"jsonrpc":"2.0","method":"call","params":{"args":[[2],["id","active","partner_id","image","__last_update","name","email","login","company_ids","company_id","companies_count","sel_groups_1_9_10","sel_groups_2_3","in_group_7","in_group_4","in_group_5","in_group_6","in_group_8","lang","tz","tz_offset","action_id","signature","display_name"]],"model":"res.users","method":"read","kwargs":{"context":{"lang":"en_US","tz":"Europe/Brussels","uid":2,"search_default_no_share":1,"bin_size":true}}},"id":212139516}),
    success: function(data){
      console.log('succes: '+data);
    },
	error: function(error) {
	console.log(error)
}
  });
```
Current behavior:
I alway get error 
```
"{"jsonrpc": "2.0", "id": 212139516, "error": {"code": 100, "message": "Odoo Session Expired", "data": {"name": "odoo.http.SessionExpiredException", "debug": "Traceback (most recent call last):\n  File \"/Users/caophuc/odoo12/odoo/http.py\", line 653, in _handle_exception\n    return super(JsonRequest, self)._handle_exception(exception)\n  File \"/Users/caophuc/odoo12/odoo/http.py\", line 312, in _handle_exception\n    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])\n  File \"/Users/caophuc/odoo12/odoo/tools/pycompat.py\", line 87, in reraise\n    raise value\n  File \"/Users/caophuc/odoo12/odoo/addons/base/models/ir_http.py\", line 192, in _dispatch\n    auth_method = cls._authenticate(func.routing[\"auth\"])\n  File \"/Users/caophuc/odoo12/odoo/addons/base/models/ir_http.py\", line 118, in _authenticate\n    getattr(cls, \"_auth_method_%s\" % auth_method)()\n  File \"/Users/caophuc/odoo12/odoo/addons/base/models/ir_http.py\", line 91, in _auth_method_user\n    raise http.SessionExpiredException(\"Session expired\")\nodoo.http.SessionExpiredException: Session expired\n", "message": "Session expired", "arguments": ["Session expired"], "exception_type": "internal_error"}}}"
```
I try with NEW database , not install any module and user i use is ```admin```. 
It work in any databse odoo 11 but can't work any database with odoo 12.
And if i use ```session_id``` in NETWORK tab of Browser. It will work.

Expected behavior:
Call API with session_id get from ```authenticate```. 

Video/Screenshot link (optional):

**Comment by pedrobaeza** [comment]
AFAIK, you are not using a proper way for interacting with the server. A good idea can be to use an intermediate library for interacting like https://github.com/OCA/odoorpc, or at least, check how it's done inside it.

**Comment by LeCaoPhuc** [comment]
Hi @pedrobaeza. 

I try https://github.com/OCA/odoorpc .. It work.. But i need use Webservice with httpRequest javascript. When test on website with ajax , Odoo 11 work but Odoo 12 is not. I think if my way wrong it  must be not work both 2 version. 

I have use a library in NPM [https://www.npmjs.com/package/nativescript-odoo](url) and it work all server and database use Odoo 11 but Odoo 12 it can't work anything and i recieve error likely ajax in browser

**Comment by pedrobaeza** [comment]
Sorry, but this is out of scope of Odoo itself. Internal things can change, and as this is not an official way of accessing Odoo (see https://www.odoo.com/documentation/12.0/webservices/odoo.html for the official one), you should adapt your code to the changes each version has. OdooRPC is working now with v12, so check how it's done.

Closing this by the reasons above.

**Comment by artaradteam** [comment]
Hi @LeCaoPhuc 
I have same problem.
did you find any solution?

**Comment by LeCaoPhuc** [comment]
Hi @artaradteam
My Solution , after you call ``authenticate``
You continue call api ``/web/session/get_session_info`` with params 
```
$.ajax({
    url: 'https://your_url/web/session/get_session_info',
    headers: {
        'Content-Type':'application/json',
    },
    method: 'POST',
    data: JSON.stringify({
 		jsonrpc: "2.0",
            	method: 'call',
		params : {}
            }),
    success: function(data){
      console.log('succes: '+data);
    },
	error: function(error) {
	console.log(error)
}
  });
```
reponse of ``get_session_info``  likely reponse of ``authenticate`` method and that is all you need.
This solution can work from odoo 10, 11, 12 with me.

Hope this help.

**Comment by artaradteam** [comment]
Hi , LePhuc Thank you for your answer.I tested your solution but I get same error.I use nginx as a web server.  do you know nginx configuration for odoo 12?thanks 

Mohammad.Hassanzadeh

Departmentof Mathematical Sciences
Sharif University of Technology
 

    On Thursday, March 7, 2019, 8:43:13 AM GMT+3:30, LePhuc <notifications@github.com> wrote:  
 
 
Hi @artaradteam
My Solution , after you call authenticate
You continue call api /web/session/get_session_info with params
$.ajax({
    url: 'https://your_url/web/session/get_session_info',
    headers: {
        'Content-Type':'application/json',
    },
    method: 'POST',
    data: JSON.stringify({
 		jsonrpc: "2.0",
            	method: 'call',
		params : {}
            }),
    success: function(data){
      console.log('succes: '+data);
    },
	error: function(error) {
	console.log(error)
}
  });

reponse of get_session_info likely reponse of authenticate method and that is all you need.
This solution can work from odoo 10, 11, 12 with me.

Hope this help.

**Comment by LeCaoPhuc** [comment]
Hi @artaradteam.
My ngnix config here: 
```
#user  nobody;
worker_processes  1;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;

    sendfile        on;


    keepalive_timeout  65;
    
    server {
        listen       80;
        server_name   yourhostname;
        add_header Strict-Transport-Security max-age=2592000;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        location / {
            proxy_pass http://yourip:8069; 
        }
    }

    server {
        listen       443 ssl;
        server_name yourhostname;
        proxy_read_timeout 720s;
        proxy_connect_timeout 720s;
        proxy_send_timeout 720s;

        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP $remote_addr;



        ssl_certificate      /etc/ssl/nginx/server.crt;
        ssl_certificate_key  /etc/ssl/nginx/server.key;
        ssl_session_timeout 30m;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;


        ssl_session_cache    shared:SSL:1m;

        ssl_ciphers  HIGH:!aNULL:!MD5;
        ssl_prefer_server_ciphers  on;

        gzip_types text/css text/scss text/plain text/xml application/xml application/json application/javascript;
        gzip on;
        location / {
            proxy_pass http://yourip:8069; 
        }
    }
    include servers/*;
}
```
You need change ``yourip`` and ``yourhostname`` . I only use it to config hostname. But when i test api it work when don't have nginx.
To ensure i think you should be use new database, because can have some cache in cookies.
You can push your code call api here to i can see.

**Comment by artaradteam** [comment]
Without nginx I get cross origin error.
What is change odoo 12 in security layer? 
I tested various types method for connection odoo 12 , but no of them work.
It work only for  'auth:none'  service.

**Comment by LeCaoPhuc** [comment]
@artaradteam 
In my case, on odoo 12 must call ``get_session_info`` to get session_id and odoo11 don't need call it ``get_session_info``.

**Comment by chaise29** [comment]
@artaradteam 
I struggled with the same problem, but I managed to find a solution there https://stackoverflow.com/questions/46288437/set-cookies-for-cross-origin-requests.

Not sure if it is specific to odoo v12, but authentication is actually made with the response header 'set-cookie : session_id=xxxxxx' that you get from the '/web/session/authenticate' url request. The 'session_id' attribute that you get from the response body is somehow not correct. 

Normally you wouldn't have to do anything, any url request you'd make after authentication should contain a 'Cookie' header containing the session_id added by your browser , but since you are calling the api from a different origin maybe you are using  the Access-Control-Allow-Origin  * wildcard on your nginx config, and authentication doesn't work in that case ...

So what you have to do in that case is to : 
- specifically allow the origin (Access-Control-Allow-Origin http://localhost:8100 in my case) 
- also make sure that Access-Control-Allow-Credentials value is set to true
- and  then, on the client side you must set the XMLHttpRequest.withCredentials flag to true ( in my case I use angular HttpClient, and it goes that way  -> 
 http.post(url, body, { headers: this.headers, withCredentials : true }) )

 Hope it helps

**Comment by artaradteam** [comment]
Hi chaise29,
Thank you for answer, by this configuration it work on browser but when I generate apk and run it on my device,it does not work,
I think value of Access-Control-Allow-Origin in device is not http://localhost:8100.
did you have any idea?
thanks

**Comment by chaise29** [comment]
I didn't try with a generated apk yet ...
I use ionic, if you are using ionic as well, you can check the documentation there -> https://ionicframework.com/docs/faq/cors#solutions-for-cors-errors
Basically , on android the origin is  http://localhost, so adding this url , without the 8100 port, should work

**Comment by quanganh206** [comment]
I do not think it is a good solution, actually it's just work on browser. Even you accept cors with http://localhost android app still not working. And specially with iOS there is no way with ionic://localhost.

**Comment by Jerther** [comment]
If that can be of any help to anyone, I just realized the JSONRPC mechanism has changed in 12.0, or at least the way we used to use it back in 8.0 has. In short, the `session_id` is not needed anymore, but the logged user ID (`uid`) has to be provided instead.

[Here is the official documentation](https://www.odoo.com/documentation/12.0/howtos/backend.html#json-rpc-library). It's in Python but it's fairly easy to translate to JS. I've been able to write a Google Apps Script that uses Odoo 12 JSONRPC which is hosted behind NGINX. I did not have to add any CORS specific config.

Also, FWIW, I believe the change occured in https://github.com/odoo/odoo/commit/9bae56acd4ebe2b169ef635892b2cd9b6664a9d0#diff-03542f2676ca18278ea62cb40f5a2261R1424

I hope that brings some light.

**Comment by sherpya** [comment]
so what is the point of having a response with a bad session_id to an authenticate api call?

**Comment by Jerther** [comment]
To my knowledge, I'd say there's no point anymore. But let's hear it from @pedrobaeza  ;)

**Comment by sherpya** [comment]
@Jerther also misleading better to not have such wrong value

**Comment by lwl021051** [comment]
I fix issue. The below is my steps.

step1: 
ajax  login the api: 'https://your_url/web/session/authenticate';

step2:
get the session id from response_header(NOT result session id);

step3:
ajax the api:  'https://your_url/web/session/get_session_info';
header set step2 response_header session id;

step4:
finish!!!

**Comment by aliomattux** [comment]
I have a question about this, I have an Angular app which uses Json RPC. I noticed Pedro say this is not the officially supported way, but i've found that xmlrpc is not very popular or supported out of box with Angular, so it would be more of an uphill battle to implement it.

I have a similar scenario where I went from Odoo 8, but now 14. In my app, the login screen will call /web/session/authenticate. The response contains no session id, or any identifiable session variable. It does return user id. When I make a request to /web/session/get_session_info with a now null session id, I get the session expired message like above, but if I make again the call with no parameters at all, it will return the data and my user id. I can also make model calls.

From this behavior, it seems that once login is called, there is no need to supply user id or session. When user is done, call logout. A bit puzzled on if this is a security problem or what would happen if multiple users are using the app at the same time, how does it track. From what i've found, there is no documentation at all describing anything about json rpc even though it is the method used by the odoo app itself.
